core/shared/src/main/scala/zio/internal/FiberContext.scala

```diff
@@ -606,9 +606,8 @@ private[zio] final class FiberContext[E, A](
    *
    * @param value The value produced by the asynchronous computation.
    */
-  private[this] final def resumeAsync: IO[E, Any] => Unit = {
-    val neverRan = new AtomicBoolean(true)
-    zio => if (neverRan.getAndSet(false) && exitAsync()) evaluateLater(zio)
+  private[this] final def resumeAsync(ref: AnyRef): IO[E, Any] => Unit = { zio =>
+    if (exitAsync(ref)) evaluateLater(zio)
```

Comment:
The old approach seemed simpler. Of course, it didn't work, but it didn't work because the `AtomicBoolean` should have been allocated in the main interpreter loop, not here. If you allocate it there, and pass it into `resumeAsync`, it will solve the problem with fewer code changes.

Reply:
I have another PR with this approach https://github.com/zio/zio/pull/1938
It fixes #1842 but not fixes #1843

Also there are more generic ABA problem related to Fiber status.
Look at following steps:
1. Starting effectAsync `A` (now in `Suspended` status)
2. Fiber got interrupted (not in `Running` status)
3. Finalizer starts another EffectAsync `B` (now in `Suspended` status)
4. some code executed callback from `A`; it will try to unsuspend fiber

Reply:
You are right. We used to have an "async counter" to deal with this problem. 

Let me take another look at the PR with this in mind...